File: paimon-core/src/main/java/org/apache/paimon/globalindex/GlobalIndexSchemaCompatibility.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.globalindex;

import org.apache.paimon.index.GlobalIndexMeta;
import org.apache.paimon.index.IndexFileMeta;
import org.apache.paimon.manifest.IndexManifestEntry;
import org.apache.paimon.table.FileStoreTable;
import org.apache.paimon.types.RowType;

import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/** Validates global index manifest entries against the current table schema. */
public final class GlobalIndexSchemaCompatibility {

    public static List<IndexManifestEntry> filterCompatible(
            FileStoreTable table, Collection<IndexManifestEntry> entries) {
        return partitionByCompatibility(table, entries).compatible();
    }

    public static CompatibilityResult partitionByCompatibility(
            FileStoreTable table, Collection<IndexManifestEntry> entries) {
        CompatibilityChecker checker = new CompatibilityChecker(table);
        List<IndexManifestEntry> compatible = new ArrayList<>();
        List<IndexManifestEntry> incompatible = new ArrayList<>();
        for (IndexManifestEntry entry : entries) {
            if (checker.isCompatible(entry.indexFile(), entry.schemaId())) {
                compatible.add(entry);
            } else {
                incompatible.add(entry);
            }
        }
        return new CompatibilityResult(compatible, incompatible);
    }

    public static List<IndexFileMeta> filterCompatibleFiles(
            FileStoreTable table, Collection<IndexFileMeta> indexFiles) {
        CompatibilityChecker checker = new CompatibilityChecker(table);
        List<IndexFileMeta> compatible = new ArrayList<>();
        for (IndexFileMeta indexFile : indexFiles) {
            if (checker.isCompatible(indexFile, indexFile.schemaId())) {
                compatible.add(indexFile);
            }
        }
        return compatible;
    }

    /** Global index manifest entries grouped by compatibility with the current table schema. */
    public static final class CompatibilityResult {

        private final List<IndexManifestEntry> compatible;
        private final List<IndexManifestEntry> incompatible;

        private CompatibilityResult(
                List<IndexManifestEntry> compatible, List<IndexManifestEntry> incompatible) {
            this.compatible = compatible;
            this.incompatible = incompatible;
        }

        public List<IndexManifestEntry> compatible() {
            return compatible;
        }

        public List<IndexManifestEntry> incompatible() {
            return incompatible;
        }
    }

    private static class CompatibilityChecker {

        private final FileStoreTable table;
        private final RowType currentRowType;
        private final Map<Long, RowType> historicalRowTypes = new HashMap<>();
        private final Set<Long> missingSchemaIds = new HashSet<>();

        private CompatibilityChecker(FileStoreTable table) {
            this.table = table;
            this.currentRowType = table.rowType();
            historicalRowTypes.put(table.schema().id(), currentRowType);
        }

        private boolean isCompatible(IndexFileMeta indexFile, Long schemaId) {
            GlobalIndexMeta globalIndex = indexFile.globalIndexMeta();
            if (globalIndex == null || schemaId == null) {
                return false;
            }

            RowType historicalRowType = historicalRowTypes.get(schemaId);
            if (historicalRowType == null && !missingSchemaIds.contains(schemaId)) {
                try {
                    historicalRowType =
                            table.schemaManager().tryGetSchema(schemaId).logicalRowType();
                    historicalRowTypes.put(schemaId, historicalRowType);
                } catch (FileNotFoundException e) {
                    missingSchemaIds.add(schemaId);
                }
            }
            return historicalRowType != null
                    && compatibleIndexedFields(globalIndex, historicalRowType, currentRowType);
        }
    }

    private static boolean compatibleIndexedFields(
            GlobalIndexMeta globalIndex, RowType historicalRowType, RowType currentRowType) {
        for (int fieldId : globalIndex.getIndexedFieldIds()) {
            if (!historicalRowType.containsField(fieldId)
                    || !currentRowType.containsField(fieldId)) {
                return false;
            }
            if (!historicalRowType
                    .getField(fieldId)
                    .type()
                    .equalsIgnoreNullable(currentRowType.getField(fieldId).type())) {
                return false;
            }
        }
        return true;
    }

    private GlobalIndexSchemaCompatibility() {}
}

File: paimon-python/pypaimon/globalindex/global_index_schema_compatibility.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

"""Validates global indexes against the current table schema."""

from typing import Collection, Dict, List, Set, Tuple

from pypaimon.schema.data_types import DataTypeParser


def filter_compatible_global_indexes(table, entries: Collection) -> List:
    """Keep entries whose indexed fields have the same logical types."""
    return partition_global_indexes_by_compatibility(table, entries)[0]


def partition_global_indexes_by_compatibility(
    table, entries: Collection
) -> Tuple[List, List]:
    """Group manifest entries by compatibility with the current schema."""
    checker = _CompatibilityChecker(table)
    compatible = []
    incompatible = []

    for entry in entries:
        target = (compatible if checker.is_compatible(
            entry.index_file, entry.schema_id) else incompatible)
        target.append(entry)
    return compatible, incompatible


def filter_compatible_global_index_files(table, index_files: Collection) -> List:
    """Keep index files compatible with the current schema."""
    checker = _CompatibilityChecker(table)
    return [
        index_file for index_file in index_files
        if checker.is_compatible(
            index_file, getattr(index_file, "schema_id", None))
    ]


class _CompatibilityChecker:

    def __init__(self, table):
        self._table = table
        current_schema = table.table_schema
        self._current_fields = _fields_by_id(current_schema.fields)
        self._historical_fields = {
            current_schema.id: self._current_fields,
        }
        self._missing_schema_ids: Set[int] = set()
        self._compatibility_cache = {}

    def is_compatible(self, index_file, schema_id) -> bool:
        global_index = index_file.global_index_meta
        if global_index is None or schema_id is None:
            return False

        fields = self._historical_fields.get(schema_id)
        if fields is None and schema_id not in self._missing_schema_ids:
            historical_schema = self._table.schema_manager.get_schema(schema_id)
            if historical_schema is None:
                self._missing_schema_ids.add(schema_id)
            else:
                fields = _fields_by_id(historical_schema.fields)
                self._historical_fields[schema_id] = fields

        if fields is None:
            return False
        field_ids = tuple(
            [global_index.index_field_id]
            + list(global_index.extra_field_ids or [])
        )
        cache_key = (schema_id, field_ids)
        if cache_key not in self._compatibility_cache:
            self._compatibility_cache[cache_key] = _compatible_indexed_fields(
                field_ids, fields, self._current_fields)
        return self._compatibility_cache[cache_key]


def _compatible_indexed_fields(field_ids, historical_fields, current_fields):
    for field_id in field_ids:
        historical_field = historical_fields.get(field_id)
        current_field = current_fields.get(field_id)
        if historical_field is None or current_field is None:
            return False
        if not _equals_ignore_nullable(
                historical_field.type, current_field.type):
            return False
    return True


def _fields_by_id(fields) -> Dict:
    return {field.id: field for field in fields}


def _equals_ignore_nullable(left, right) -> bool:
    left_copy = DataTypeParser.parse_data_type(left.to_dict())
    right_copy = DataTypeParser.parse_data_type(right.to_dict())
    left_copy.nullable = True
    right_copy.nullable = True
    return left_copy == right_copy
